seabios refactors

AndersonTorres opened the pull request:
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/342650

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron reviewed (commented):
Is the filename of the non-CSM build still `Csm16.bin`? I've sort of assumed that for #345324.

AndersonTorres commented:
No, it's not. I need to fix this.

AndersonTorres commented:
@SigmaSquadron please look and test it.

AndersonTorres commented:
@ofborg build seabios-qemu seabios-coreboot

SigmaSquadron commented:
I'd like to see the commit split into multiple ones, if possible: one for formatting, another for the biosfile stuff, and another for passthru. As it stands the diff is kind of hard to review.

AndersonTorres committed:
seabios: nixfmt-rfc-style

AndersonTorres committed:
seabios: use outputDoc

AndersonTorres committed:
seabios: new parameter ___build-type

It can assume the values "coreboot", "csm" and "qemu".

The triple-underscored kebab-cased format is a poor man's mitigation for the
problem of unexpected shadowing.

AndersonTorres committed:
seabios: passthru attribute `firmware`

So that the caller can pick the exact location of the resulting BIOS.

AndersonTorres committed:
{seabios-coreboot, seabios-csm, seabios-qemu}: "init"

AndersonTorres commented:
> I'd like to see the commit split into multiple ones, if possible: one for formatting, another for the biosfile stuff, and another for passthru. As it stands the diff is kind of hard to review.

Done!

SigmaSquadron commented on pkgs/by-name/se/seabios/package.nix:
```diff
@@ -1,11 +1,37 @@
+  stdenv,
+  writeText,
+  # Configurable options
+  ___build-type ? "csm",
```
```suggestion
  buildType ? "csm",
```
I don't see a reason for the underline-prefixed, kebab-case name.

AndersonTorres replied:
Paranoia:

https://github.com/NixOS/nixpkgs/pull/324199#issuecomment-2212422318

there is a small, non-negligible chance of a package having the name `buildType`.

SigmaSquadron replied:
Fascinating. I actually got a similar thing when I was working on `starsector`. Leave it as is then.

SigmaSquadron commented on pkgs/by-name/se/seabios/package.nix:
```diff
@@ -29,31 +58,42 @@ stdenv.mkDerivation (finalAttrs: {
   '';
 
+  passthru = {
+    build-type = ___build-type;
```
```suggestion
    inherit buildType;
```

SigmaSquadron commented on pkgs/by-name/se/seabios/package.nix:
```diff
@@ -64,9 +104,7 @@ stdenv.mkDerivation (finalAttrs: {
-    platforms = lib.systems.inspect.patternLogicalAnd
-      lib.systems.inspect.patterns.isUnix
-      lib.systems.inspect.patterns.isx86;
+    platforms = lib.systems.inspect.patternLogicalAnd lib.systems.inspect.patterns.isUnix lib.systems.inspect.patterns.isx86;
```
```suggestion
    platforms = lib.intersectLists lib.platforms.unix lib.platforms.x86_64;
```

AndersonTorres replied:
String-y platforms are considered legacy (https://github.com/NixOS/nixpkgs/blob/fd698a4ab779fb7fb95425f1b56974ba9c2fa16c/lib/meta.nix#L208)

SigmaSquadron replied:
Oh wow. Why is everyone still using it?

AndersonTorres replied:
I believe it is because the modern way is not so well documented and it is a bit harder to understand.

Further, there are some dark corners on Nixpkgs that still expect the stringy values.

SigmaSquadron reviewed (commented):
Overall looks good; a few nits:

SigmaSquadron commented on pkgs/by-name/se/seabios/package.nix:
```diff
@@ -29,31 +58,42 @@ stdenv.mkDerivation (finalAttrs: {
 
+  passthru = {
+    build-type = ___build-type;
+    firmware-location = "${finalAttrs.finalPackage}/share/seabios/${biosfile}";
```
```suggestion
    firmware = "${finalAttrs.finalPackage}/share/seabios/${biosfile}";
```
For parity with OVMF.

AndersonTorres replied:
OK!

SigmaSquadron reviewed (approved):
Sorry for bothering you with the spurious reviews, and thank you for addressing this so quickly! LGTM.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2009
